pkgs/development/tools/ocaml/camlp5/default.nix

```
@@ -14,92 +14,85 @@
   legacy ? false,
 }:
 
-if lib.versionOlder ocaml.version "4.02" then
-  throw "camlp5 is not available for OCaml ${ocaml.version}"
-else
-
+stdenv.mkDerivation (
+  finalAttrs:
   let
-    params =
-      if lib.versionAtLeast ocaml.version "4.12" && !legacy then
-        rec {
-          version = "8.03.02";
+    recent = lib.versionAtLeast (lib.versions.major finalAttrs.version) "8";
+  in
+  {
 
-          src = fetchFromGitHub {
-            owner = "camlp5";
-            repo = "camlp5";
-            rev = version;
-            hash = "sha256-nz+VfGR/6FdBvMzPPpVpviAXXBWNqM3Ora96Yzx964o=";
-          };
+    version = if lib.versionAtLeast ocaml.version "4.12" && !legacy then "8.04.00" else "7.14";
 
-          nativeBuildInputs = [
-            makeWrapper
-            ocaml
-            findlib
-            perl
-          ];
-          buildInputs = [
-            bos
-            pcre2
-            re
-            rresult
-          ];
-          propagatedBuildInputs = [ camlp-streams ];
-          postInstall = ''
-            for prog in camlp5 camlp5o camlp5r camlp5sch mkcamlp5 ocpp5
-            do
-              wrapProgram $out/bin/$prog \
-                --prefix CAML_LD_LIBRARY_PATH : "$CAML_LD_LIBRARY_PATH"
-            done
-          '';
+    pname = "ocaml${ocaml.version}-camlp5";
 
+    src = fetchFromGitHub {
+      owner = "camlp5";
+      repo = "camlp5";
+      tag =
+        if recent then
+          finalAttrs.version
+        else
+          "rel${builtins.replaceStrings [ "." ] [ "" ] finalAttrs.version}";
+      hash =
+        {
+          "8.04.00" = "sha256-5IQVGm/tqEzXmZmSYGbGqX+KN9nQLQgw+sBP+F2keXo=";
+          "8.03.2" = "sha256-nz+VfGR/6FdBvMzPPpVpviAXXBWNqM3Ora96Yzx964o=";
+          "7.14" = "sha256-/ORtS0uc/GN+g3y6N5ftjL4OBSqV6iswLRbfpeNCprU=";
         }
-      else
-        rec {
-          version = "7.14";
-          src = fetchFromGitHub {
-            owner = "camlp5";
-            repo = "camlp5";
-            rev = "rel${builtins.replaceStrings [ "." ] [ "" ] version}";
-            sha256 = "1dd68bisbpqn5lq2pslm582hxglcxnbkgfkwhdz67z4w9d5nvr7w";
-          };
-          nativeBuildInputs = [
-            ocaml
-            perl
-          ];
-        };
-  in
+        ."${finalAttrs.version}";
+    };
+    nativeBuildInputs = [
+      ocaml
+      perl
+    ]
+    ++ lib.optionals recent [
+      makeWrapper
+      findlib
+    ];
 
-  stdenv.mkDerivation (
-    params
-    // {
+    buildInputs = lib.optionals recent [
+      bos
+      pcre2
+      re
+      rresult
+    ];
 
-      pname = "ocaml${ocaml.version}-camlp5";
+    propagatedBuildInputs = lib.optional recent camlp-streams;
```

cosmetic, i think this would be better (feel free to dismiss)

```suggestion
    propagatedBuildInputs = lib.optionals recent [
      camlp-streams
    ];
```